Sortmerna plugin

smruthi98 opened the pull request:
- created sortmerna.py to run sortmerna
- included unittests
- mock databases and index created 
- updated travis.yml to load sormerna and export db as an environment variable

smruthi98 committed:
Create DB

smruthi98 committed:
1. README.md - Included documentation for SortMeRNA
2. Sortmerna.py - Filtering ribosomal reads from metaranscriptomic samples.Generates 2 artifacts:ribsomal and non-ribsomal reads for each sample passed
3. test_sortmerna.py - Unittest for sortmerna.py.Current SortMeRNA version(v2.1) does not accept compressed (.gz) files. Databases and index  are stored under sortmerna/databases

Databases and index files for rRNA for unittest
Adding QC_SORTMERNA_DB_DP as environ variable
Included cython installation in travis.yml

smruthi98 commented:
Testing plugin using travis since I do not currently have a qp-shogun environment to test with

ElDeveloper commented:
> Testing plugin using travis since I do not currently have a qp-shogun environment to test with

@smruthi98 feel free to ping me with this PR is ready for a review.

smruthi98 committed:
Fixed minor linting issues

antgonza commented on .travis.yml:
```diff
@@ -13,8 +13,9 @@ before_install:
-  - travis_retry conda create -q --yes -n qiita python=3.6 pip libgfortran numpy nginx
+  - travis_retry conda create -q --yes -n qiita python=3.6 pip libgfortran numpy nginx 
   - source activate qiita
+  - pip install cython
```
Suggest adding this to the conda create line 16

smruthi98 replied:
Done.

antgonza commented on qp_shogun/sortmerna/__init__.py:
```diff
@@ -0,0 +1,18 @@
+
+# Defining the Sortmerna command
+req_params = {'input': ('artifact', ['per_sample_FASTQ'])}
+opt_params = {
```
Couldn't add in the actual files but can you remove _all_ the .DS_Store from this PR and add it to the .gitignore file?

smruthi98 replied:
Added

smruthi98 replied:
Done.

antgonza commented on qp_shogun/sortmerna/__init__.py:
```diff
@@ -0,0 +1,18 @@
+req_params = {'input': ('artifact', ['per_sample_FASTQ'])}
+opt_params = {
+    # Threads used
+    'Number of threads': ['integer', '5']
```
Could you add the other SMR parameters, even if using defaults?

smruthi98 replied:
Added.

antgonza commented on qp_shogun/sortmerna/databases/rRNA_databases/scripts/generate_databases.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
The fasta files are pretty big, perhaps worth deleting them and adding their gz version?

smruthi98 replied:
Compressed files.

antgonza commented on qp_shogun/sortmerna/databases/rRNA_databases/scripts/generate_databases.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+rootdir="/home/skarthikeyan/ARB_SILVA_databases"
```
Is this file actually being used?

smruthi98 replied:
File was removed

antgonza commented on qp_shogun/sortmerna/databases/rRNA_databases/scripts/generate_databases.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+rootdir="/home/skarthikeyan/ARB_SILVA_databases"
```
perhaps remove this folder and add a comment to add your own?

smruthi98 replied:
Done.

antgonza commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+# ).format(DIR)
+
+SORTMERNA_PARAMS = {
+    'a': 'Number of threads'}
```
Just to also have it here; it will be good to expose all SRN parameters ...

smruthi98 replied:
Added

antgonza commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+        if r_fp is None:
+            cmds.append('sortmerna --ref %s --reads %s '
+                        '--aligned %s --other %s '
+                        '--fastx --blast 1 --num_alignments 1 -a %s' % (
```
yeah, it will be good to expose this parameters, in case we need to change them in the future; even if the user can't change them

smruthi98 replied:
Modified accordingly.

antgonza commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+                            join(out_dir, '%s.nonribosomal.R2' % run_prefix),
+                            threads))
+    return cmds, samples
+    # In this version I have not added a summary file or sam file
```
can you open an issue and list it here?

smruthi98 replied:
#64 : opened issue

antgonza commented on qp_shogun/sortmerna/tests/test_sortmerna.py:
```diff
@@ -0,0 +1,192 @@
+
+    def setUp(self):
+        plugin("https://localhost:21174", 'register', 'ignored')
+        # db_path = os.environ["QC_SORTMERNA_DB_DP"]
```
if not needed, can you remove?

antgonza commented on qp_shogun/sortmerna/tests/test_sortmerna.py:
```diff
@@ -0,0 +1,192 @@
+    def setUp(self):
+        plugin("https://localhost:21174", 'register', 'ignored')
+        # db_path = os.environ["QC_SORTMERNA_DB_DP"]
+        self.params = {
```
also for completeness, adding here, can you add all parameters, thank you!

smruthi98 replied:
Added.

antgonza commented on qp_shogun/sortmerna/tests/test_sortmerna.py:
```diff
@@ -0,0 +1,192 @@
+
+        fp1_1 = join(in_dir, 'kd_test_1_R1.fastq')
+        fp1_2 = join(in_dir, 'kd_test_1_R2.fastq')
+        # fp2_1 = join(in_dir, 'kd_test_2_R1.fastq.gz')
```
can you rm lines if not used?

antgonza commented on qp_shogun/sortmerna/tests/test_sortmerna.py:
```diff
@@ -0,0 +1,192 @@
+        # copyfile('support_files/kd_test_1_R2.fastq.gz', fp2_2)
+        # inserting new prep template
+        prep_info_dict = {
+            # 'SKB7.640196': {'run_prefix': 'kd_test_1'},
```
rm if not needed.

antgonza commented on qp_shogun/sortmerna/utils.py:
```diff
@@ -0,0 +1,177 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
```
This is a copy from other file, right? If that's correct, could you simply import from that other package vs. copy the full file?

smruthi98 replied:
The one under sortmerna has a slightly modified version to account for the uncompressed output files from the Sortmerna command.

antgonza reviewed (commented):
@smruthi98, thank you for putting this PR together, it's getting there. A few comments.

ElDeveloper commented on setup.py:
```diff
@@ -43,7 +43,8 @@
       packages=['qp_shogun', 'qp_shogun/trim',
-                'qp_shogun/filter', 'qp_shogun/shogun'],
+                'qp_shogun/filter', 'qp_shogun/shogun',
+                'qp_shogun/sortmerna'],
```
If we need to keep the data in the `databases` folder, we'll need to add an entry to the package_data attribute to include those files.

smruthi98 replied:
I've compressed all the database files for the time being (can always work on a longer term solution in the meanwhile). Modified setup.py accordingly.

ElDeveloper commented on qp_shogun/sortmerna/README.md:
```diff
@@ -0,0 +1,134 @@
+# rRNA removal
+
+## SortMeRNA for MetaT analysis
```
```suggestion
## SortMeRNA for Metatranscriptomics analysis
```

ElDeveloper commented on qp_shogun/sortmerna/README.md:
```diff
@@ -0,0 +1,134 @@
+
+Updated  04/07/2020
+
+SortMeRNA for removal of ribsomal reads from quality filtered metaT data
```
```suggestion
SortMeRNA for removal of ribsomal reads from quality filtered metatranscriptomics data.
```

ElDeveloper commented on qp_shogun/sortmerna/README.md:
```diff
@@ -0,0 +1,134 @@
+residual ribosomal RNA in the libraries that you want to remove/separate from
+the non ribosomal RNA sequences.
+
+Input: Quality filtered metaT reads (Fasta/Fastq) 
```
```suggestion
Input: Quality filtered metatranscriptomics reads (FASTA/FASTQ) 
```

ElDeveloper commented on qp_shogun/sortmerna/README.md:
````diff
@@ -0,0 +1,134 @@
+
+## Indexing custom databases
+
+```./indexdb_rna --ref db.fasta,db.idx ```
````
What is `indexdb_rna`?

smruthi98 replied:
Only required if building custom databases as databases need to be indexed before running.

ElDeveloper replied:
Thanks, although the command call would be without `./`, right?

```suggestion
```indexdb_rna --ref db.fasta,db.idx ```
```

ElDeveloper commented on qp_shogun/sortmerna/README.md:
```diff
@@ -0,0 +1,134 @@
+
+## Currently indexed databases 
+
+rRNA databases are downloaded and stored at /opt/genome/qiita_dbs/sortmerna/rna-depletion/rRNA_databases. 
+
+Downloaded databases have been indexed and indexes are stored at /opt/genome/qiita_dbs/sortmerna/rna-depletion/index.
+
+rRNA from bacteria, archaea, eukarya (SILVA and rfam database)
```
@antgonza  is this sort of documentation needed for a plugin?

antgonza replied:
Good question! I think it will be great to have it in the main Qiita documentation only (which AFAIK) we already have as this will help avoid duplication and having users to search and find it within the plugins.

smruthi98 replied:
Removed from the Plugin README version

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+from os.path import join
```
Can you include the copyright header:

```
 # -----------------------------------------------------------------------------
 # Copyright (c) 2014--, The Qiita Development Team.
 #
 # Distributed under the terms of the BSD 3-clause License.
 #
 # The full license is in the file LICENSE, distributed with this software.
 # -----------------------------------------------------------------------------
```

smruthi98 replied:
Included.

ElDeveloper commented on qp_shogun/sortmerna/__init__.py:
```diff
@@ -0,0 +1,18 @@
+from qiita_client import QiitaCommand
```
Can you include the copyright header:

```
 # -----------------------------------------------------------------------------
 # Copyright (c) 2014--, The Qiita Development Team.
 #
 # Distributed under the terms of the BSD 3-clause License.
 #
 # The full license is in the file LICENSE, distributed with this software.
 # -----------------------------------------------------------------------------
```

smruthi98 replied:
Inlcuded

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+
+def generate_sortmerna_commands(forward_seqs, reverse_seqs, map_file,
+                                out_dir, parameters):
+    """Generates the Sortmerna commands
+    Parameters
```
```suggestion
    """Generates the Sortmerna commands
    
    Parameters
```

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+    out_dir : str
+        The job output directory
+    parameters : dict
+        The command's parameters, keyed by parameter name
+    Returns
```
```suggestion
        The command's parameters, keyed by parameter name

    Returns
```

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+    -------
+    cmds: list of str
+        The Sortmerna commands
+    samples: list of tup
+        list of 4-tuples with run prefix, sample name, fwd read fp, rev read fp
+    Notes
```
```suggestion
        list of 4-tuples with run prefix, sample name, fwd read fp, rev read fp

    Notes
```

smruthi98 replied:
Changed

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+    samples = make_read_pairs_per_sample(forward_seqs, reverse_seqs, map_file)
+
+    cmds = []
+    # param_string = _format_params(parameters, SORTMERNA_PARAMS)
```
```suggestion
```

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+
+    threads = parameters['Number of threads']
+
+    for run_prefix, sample, f_fp, r_fp in samples:
+        if r_fp is None:
+            cmds.append('sortmerna --ref %s --reads %s '
+                        '--aligned %s --other %s '
+                        '--fastx --blast 1 --num_alignments 1 -a %s' % (
+                            RNA_REF_DB, f_fp,
+                            join(out_dir, '%s.ribosomal.R1' % run_prefix),
+                            join(out_dir, '%s.nonribosomal.R1' % run_prefix),
+                            threads))
+        else:  # there is a forward and reverese seq
+            cmds.append('sortmerna --ref %s --reads %s '
+                        '--aligned %s --other %s '
+                        '--fastx --blast 1 --num_alignments 1 -a %s' % (
+                            RNA_REF_DB, f_fp,
+                            join(out_dir, '%s.ribosomal.R1' % run_prefix),
+                            join(out_dir, '%s.nonribosomal.R1' % run_prefix),
+                            threads))
+
```
I think the forward read command can be added always, like:

```suggestion
    for run_prefix, sample, f_fp, r_fp in samples:
        cmds.append('sortmerna --ref %s --reads %s '
                    '--aligned %s --other %s '
                    '--fastx --blast 1 --num_alignments 1 -a %s' % (
                        RNA_REF_DB, f_fp,
                        join(out_dir, '%s.ribosomal.R1' % run_prefix),
                        join(out_dir, '%s.nonribosomal.R1' % run_prefix),
                        threads))
        # there is a forward and reverese seq
        if r_fp is not None:
```

smruthi98 replied:
Yep, makes sense. Modified accordingly.

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+
+
+def sortmerna(qclient, job_id, parameters, out_dir):
+    """Run Sortmerna with the given parameters
+    Parameters
+    ----------
+    qclient : tgp.qiita_client.QiitaClient
+        The Qiita server client
+    job_id : str
+        The job id
+    parameters : dict
+        The parameter values
+    out_dir : str
+        The path to the job's output directory
+    Returns
```
```suggestion
    """Run Sortmerna with the given parameters

    Parameters
    ----------
    qclient : tgp.qiita_client.QiitaClient
        The Qiita server client
    job_id : str
        The job id
    parameters : dict
        The parameter values
    out_dir : str
        The path to the job's output directory

    Returns
```

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+    fps = artifact_info['files']
+
+    # Get the artifact metadata
+
```
```suggestion
```

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+    # reads and other for the non-ribosomal reads
+    # Step 4 generating artifacts for Nonribosomal reads
+
+    msg = "Step 4 of 5: Generating artifacts "
+    "for Nonribosomal reads (%d/{0})".format(len_cmd)
```
I'm surprised this was not flagged by flake8.

```suggestion
    msg = ("Step 4 of 5: Generating artifacts "
           "for Nonribosomal reads (%d/{0})").format(len_cmd)
```

ElDeveloper commented on qp_shogun/sortmerna/sortmerna.py:
```diff
@@ -0,0 +1,180 @@
+    ainfo.append(_per_sample_ainfo(
+        out_dir, samples, suffixes, prg_name, file_type_name, bool(rs)))
+    # Step 5 generating artifacts for Ribosomal reads
+    msg = "Step 5 of 5: Generating artifacts "
+    "for Ribosomal reads (%d/{0})".format(len_cmd)
```
```suggestion
    msg = ("Step 5 of 5: Generating artifacts "
           "for Ribosomal reads (%d/{0})").format(len_cmd)

```

ElDeveloper commented on qp_shogun/sortmerna/utils.py:
```diff
@@ -0,0 +1,177 @@
+    At this stage it is required that if reverse sequences are present that all
+    samples have both a forward and a reverse sequence. However, the read
+    trimming/filtering step can sometimes eliminate all reverse reads,
+    especially in low
+    coverage samples with poor overall reverse read quality.
```
```suggestion
    especially in low coverage samples with poor overall reverse read quality.
```

ElDeveloper commented on qp_shogun/sortmerna/utils.py:
```diff
@@ -0,0 +1,177 @@
+    # check that rev seqs are same len
+    if reverse_seqs:
+        if len(forward_seqs) != len(reverse_seqs):
+            raise ValueError('Your reverse and forward files are of different '
+                             'length. Forward: %s. Reverse: %s.' %
```
The word length can be confusing in the context of sequence files, I would suggest something like:

```suggestion
            raise ValueError('There's a mistmatching count of forward and reverse files.'
                             'Forward: %s. Reverse: %s.' %
```

ElDeveloper commented on qp_shogun/sortmerna/utils.py:
```diff
@@ -0,0 +1,177 @@
+            if fwd_fn.startswith(rp) and run_prefix is None:
+                run_prefix = rp
+            elif fwd_fn.startswith(rp) and run_prefix is not None:
+                raise ValueError('Multiple run prefixes match this fwd read: '
+                                 '%s' % fwd_fn)
```
I would suggest also mentioning what the problematic run prefixes are.
```suggestion
                raise ValueError('Multiple run prefixes match this fwd read: '
                                 '%s. Matches are: %s and %s' % (fwd_fn, rp, run_prefix))
```

ElDeveloper reviewed (changes requested):
Thanks for putting this together @smruthi98! Like @antgonza mentioned, there's a few `.DS_Store` files checked-in. Can you remove those files from your branch? Do we need to have the files in `qp_shogun/sortmerna/databases/` or could we download these as part of the Travis deploy?

smruthi98 committed:
Update qp_shogun/sortmerna/README.md

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/README.md

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/sortmerna.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/utils.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/utils.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/utils.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/sortmerna.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Update qp_shogun/sortmerna/sortmerna.py

Co-Authored-By: Yoshiki Vázquez Baeza <yoshiki@ucsd.edu>

smruthi98 committed:
Compressed DB, updated documentation and made the suggested changes

Updated travis.yml to include cython and uncompressing db files
Compressed databases to .tar.bz2 format
Removed .DS_Store files
Included params

smruthi98 committed:
Merge branch 'sortmerna' of https://github.com/smruthi98/qp-shogun into sortmerna

smruthi98 committed:
Update .gitignore

smruthi98 committed:
Delete .DS_Store

smruthi98 committed:
Delete .DS_Store

smruthi98 committed:
Delete .DS_Store

smruthi98 committed:
Delete .DS_Store

smruthi98 committed:
updated cython in travis.yml

smruthi98 committed:
Merge branch 'sortmerna' of https://github.com/smruthi98/qp-shogun into sortmerna

smruthi98 committed:
Update .travis.yml

smruthi98 committed:
Update .gitignore

smruthi98 committed:
Update test_sortmerna.py

smruthi98 committed:
Added conda-forge channel in line 16

build failed at libgfortran earlier

smruthi98 committed:
Compressed index files and removed extras

smruthi98 committed:
Update .travis.yml

smruthi98 committed:
updatd params

smruthi98 committed:
fixed sytanx error

smruthi98 committed:
fixed linting issues

smruthi98 committed:
Update __init__.py

smruthi98 committed:
modified format_params

smruthi98 committed:
updating param order to match sorted list

smruthi98 committed:
removed additional space in params and removed unused funcs in sortmerna/utils.py

smruthi98 committed:
Update test_sortmerna.py

ElDeveloper reviewed (approved):
Looks good, thanks @smruthi98. Just one comment.

smruthi98 committed:
Updating to latest databases

smruthi98 committed:
DB update for new version

Updated databses to reflect the most recent, can also be used in case we decide to switch to the newer SMR version

ElDeveloper commented:
Thanks so much @smruthi98! 🎉
